### PR TITLE
[V2] sgx_main: clean up sgx_drv_probe()

### PR DESCRIPTION
**Pull request by jiazhang0**
It is unnecesary to implement the same logic in two places -
sgx_init_platform() and sgx_drv_probe().

One bad example is that the evaluation of max enclave size is wrong
in sgx_drv_probe(). Due to the calling order, the final result is
still correct.

Therefore, move all sanity checks into the internal function
sgx_init_platform() which needs to know all details of sgx.

Signed-off-by: Jia Zhang <qianyue.zj@alibaba-inc.com>

### CODE REVIEW

File: sgx_main.c [hunk 1]
```diff
@@ -377,10 +377,10 @@ static int sgx_drv_probe(struct platform_device *pdev)
 
 	cpuid_count(SGX_CPUID, 0x0, &eax, &ebx, &ecx, &edx);
 	if (edx & 0xFFFF) {
+		sgx_encl_size_max_32 = 2ULL << (edx & 0xFF);
```
Comment (light1021) [inline sgx_main.c]: looks good

**Review by light1021 — COMMENTED** [review]
I think this change looks good.

**Comment by sergeay** [comment]
No this change has some issue.
A part of the removed logic in sgx_drv_probe() is necessary, in particular the atomic operation on sgx_init_flag.
Do not merge this change and wait for pull request 38 to be merged.

**Comment by jiazhang0** [comment]
@sergeay Thanks for your comments!

I updated this PR. Here is quick diff against V1:
```
diff --git a/sgx_main.c b/sgx_main.c
index d7eb2fb..ea32be8 100644
--- a/sgx_main.c
+++ b/sgx_main.c
@@ -344,6 +344,11 @@ static atomic_t sgx_init_flag = ATOMIC_INIT(0);
 
 static int sgx_drv_probe(struct platform_device *pdev)
 {
+       if (atomic_cmpxchg(&sgx_init_flag, 0, 1)) {
+               pr_warn("intel_sgx: second initialization call skipped\n");
+               return 0;
+       }
+
        return sgx_dev_init(&pdev->dev);
 }
```

**Comment by sergeay** [comment]
Anyway there is another pull request (#38) in parallel to yours that includes your changes and more.
So please wait until this request is merged. It  is right under review and testing.
Thanks.

**Pull request closed without merging**